docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,602 @@
+gadgets approach that categorization is softer, each gadget defines how it behaves according to the
+maps / types they use and it's possible that the same gadget implements different behaviors, like
+tracer and topper, etc.
+
+### Maps
+
+#### `RingBuf` and `PerfEventArray` with `print_` Prefix (a.k.a tracers)
+
+These maps indicate that the gadget produces a stream of events. Inspektor Gadget uses the BTF
+information to format and print the events.
+
+```c
+// taken from gadgets/trace_open.bpf.c
+struct event {
+	__u64 timestamp;
+	__u32 pid;
+	__u32 uid;
+	__u32 gid;
+	mnt_ns_id_t mntns_id;
+	int ret;
+	int flags;
+	__u16 mode;
+	__u8 comm[TASK_COMM_LEN];
+	__u8 fname[NAME_MAX];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__type(value, struct event);
+} print_events SEC(".maps");
+```
+
+#### `HashMap` with `stats_` Prefix (a.k.a toppers)
+
+These maps are used to implement toppers, i.e. gadgets that print a list of elements sorted by
+specific parameters, like TCP connections by sent bytes.
+
+Inspektor Gadget automatically reads and sorts those maps according to the configuration provided by the user:
+- Interval: How often to print the map contents
+- Sort by: Field within the event type to sort by
+
+#### BPF Iterators (a.k.a snapshotters)
+
+Programs of type `iter/` are automatically loaded and attached by Inspektor Gadget, then they are
+triggered and its output is parsed according to the BTF information provided by it.
+
+In order to have BTF information available, the programs have to define an unused variable with the
+type they output in the iterator:
+
+```c
+const volatile event_t myiterator_output = {};
+
+const struct myevent *gadget_iter_type __attribute__((unused));
+
+SEC("iter/task_file")
+int myiterator(struct bpf_iter__task_file *ctx)
+```
+
+#### `HashMap` with `hist_` Prefix (a.k.a profilers)
+
+These maps are use to implement profilers: gadgets that output a histogram. The histogram collection
+starts when the gadget is run, and it's printed when the gadgets stops.
+
+TODO1: It's possible to also print the histogram periodically, however it's not explored yet.
+
+TODO2: There is probably some overlap with histogram metrics support for Prometheus. It's very
+likely that the same gadget can be used for both purposes.
+
+#### Prometheus
+
+TODO: We can heavily base this on https://github.com/cloudflare/ebpf_exporter.
+
+##### Counters
+
+##### Histograms
+
+##### Gauges
+
+
+### Inspektor Gadget API for Gadgets Developers
+
+Inspektor Gadget provides some C header defining some types and functions to be used by gadgets.
+Those types and functions are used to enable some of the features described below.
+
+#### Data Types and Helper Headers
+
+Types provided in pkg/gadgets/common/types.h are used to format data in a specific way.
+
+- `ip_addr`: Union representing either IPv4 or IPv6 address.
```

Comment:
I think it can be useful when not enrichment is desired for a given IP address (because of performance reasons). However I think we can go ahead with only the endpoints and introduce other types if needed later on. I'll remove `ip_addr` then.

Reply:
> I think it can be useful when not enrichment is desired for a given IP address (because of performance reasons).

I think this should be configurable by the user of the gadget and not the author. We need to implement "toggleable" operators anyway

Reply:
Yeah, I agree. We can try to enrich all IPs, but providing an option for the user to disable it.